File: app/admin/rank.py
"""The "Fertigstellen" traffic-light rank.

A spot's rank summarises how far along its rework is, from its readiness gaps:

* **green**  — nothing open (no required gaps left).
* **yellow** — one or two points open, and the hero image is present.
* **red**    — an *important* input is missing (hero image) OR more than two
               points are open.

Climatology is a *soft* gap: it runs as a background job (auto-started on
go-live), not a manual omission, so a pending climatology only ever keeps a spot
yellow — it never forces red and never counts toward the "more than two" red
threshold.

Operators may override the automatic value per spot (``spots.finish_rank``);
the override wins until cleared. Keep this rule in sync with the frontend copy
in ``frontend/src/lib/rank.ts``.
"""

from __future__ import annotations

# Gaps that count as "important" — their absence forces red on their own.
IMPORTANT_GAPS = {"image"}
# Gaps that must never push a spot to red (kept yellow at worst).
SOFT_GAPS = {"climatology"}
RANKS = ("red", "yellow", "green")


def auto_rank(gaps: list[str]) -> str:
    """Traffic-light rank derived purely from the readiness ``gaps`` list."""
    if not gaps:
        return "green"
    hard = [g for g in gaps if g not in SOFT_GAPS]
    if any(g in IMPORTANT_GAPS for g in gaps) or len(hard) > 2:
        return "red"
    return "yellow"


def effective_rank(gaps: list[str], override: str | None) -> str:
    """The rank actually shown: a valid manual override, else the auto value."""
    if override in RANKS:
        return override
    return auto_rank(gaps)

File: frontend/src/lib/rank.ts
// The "Fertigstellen" traffic-light rank, mirrored from app/admin/rank.py.
//   green  — nothing open
//   yellow — one or two points open (hero image present)
//   red    — hero image missing OR more than two points open
// Climatology is a soft gap (auto-started on go-live): it only ever keeps a spot
// yellow, never forces red. An operator override wins over the automatic value.

import type { Rank } from "./api";

// Gaps that count as "important" — their absence forces red on their own.
const IMPORTANT_GAPS = new Set(["image"]);
// Gaps that must never push a spot to red (kept yellow at worst).
const SOFT_GAPS = new Set(["climatology"]);

export const RANKS: Rank[] = ["green", "yellow", "red"];

export function autoRank(gaps: string[]): Rank {
  if (gaps.length === 0) return "green";
  const hard = gaps.filter((g) => !SOFT_GAPS.has(g));
  if (gaps.some((g) => IMPORTANT_GAPS.has(g)) || hard.length > 2) return "red";
  return "yellow";
}

export function effectiveRank(gaps: string[], override: Rank | null): Rank {
  return override ?? autoRank(gaps);
}

export const RANK_LABEL: Record<Rank, string> = {
  red: "Rot",
  yellow: "Gelb",
  green: "Grün",
};

// Solid dot colour for a rank. Uses the admin semantic tokens (theme-aware:
// legible in both light and dark) — rank.ts is an admin-only helper.
export const RANK_DOT: Record<Rank, string> = {
  red: "bg-admin-danger",
  yellow: "bg-admin-warning",
  green: "bg-admin-success",
};

// Tinted card styling (border + background) for a ranked row. Admin semantic
// token surfaces so the traffic-light tint adapts to light/dark.
export const RANK_CARD: Record<Rank, string> = {
  red: "border-admin-danger-border bg-admin-danger-bg",
  yellow: "border-admin-warning-bg bg-admin-warning-bg",
  green: "border-admin-success-border bg-admin-success-bg",
};
